### PR TITLE
RUN-3900:  Hashi corp vault integration modification time issue

### PR DESCRIPTION
**Pull request by edbaltra**
JIRA TICKET: https://pagerduty.atlassian.net/browse/RUN-3327

This pull request adds support for reading and using Vault metadata (such as creation and update timestamps) from HashiCorp Vault KV v2 secrets, and ensures these timestamps are set correctly on resources. It also introduces comprehensive unit tests to verify correct and robust timestamp handling, including various edge cases.

**Vault metadata extraction and timestamp handling:**

* Added a `vaultMetadata` field (with getter and setter) to the `KeyObject` class to store Vault metadata, and included it in the `toString()` method for better debugging. [[1]](diffhunk://#diff-0a219caee68219df19c62cdfc3e559a18b3264a0315fd094be7a20efff245882R18) [[2]](diffhunk://#diff-0a219caee68219df19c62cdfc3e559a18b3264a0315fd094be7a20efff245882R88-R95) [[3]](diffhunk://#diff-0a219caee68219df19c62cdfc3e559a18b3264a0315fd094be7a20efff245882R104)
* In `KeyObjectBuilder`, extracted Vault KV v2 metadata from `LogicalResponse` and set it on `VaultKey` instances, both for regular and parent objects. [[1]](diffhunk://#diff-d982155eaa7f195e76b3ebf1fdffa8afe7273a8781e1f7ff09f241f2f65458beR50-R63) [[2]](diffhunk://#diff-d982155eaa7f195e76b3ebf1fdffa8afe7273a8781e1f7ff09f241f2f65458beL100-R119)
* In `VaultKey`, parsed `created_time` and `updated_time` from Vault metadata (RFC3339 format), set them as resource creation and modification times, and handled missing or invalid timestamps gracefully with logging. [[1]](diffhunk://#diff-c089db614dd84a61566421c4d789dbe6e055bca1489c813a34ae7fbbd8a3593bR17-R32) [[2]](diffhunk://#diff-c089db614dd84a61566421c4d789dbe6e055bca1489c813a34ae7fbbd8a3593bR162-R195)

**Testing and validation:**

* Added a new test class `VaultKeyTest` with thorough unit tests covering correct parsing of timestamps, fallback logic, handling of missing/invalid metadata, and correct resource creation for both password and private key content types.

### CODE REVIEW

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java [hunk 1]
```diff
@@ -45,10 +47,18 @@ KeyObject build(){
...
+            if (dataMetadata != null && !dataMetadata.isEmpty()) {
+                metadata = dataMetadata.getMetadataMap();
+            }
+
             if(data !=null) {
                 object = new RundeckKey(response,path);
             }else{
                 object = new VaultKey(response,path);
+                object.setVaultMetadata(metadata);
             }
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java:64]: RundeckKey doesn't need Vault metadata because it stores timestamps directly in its payload (lines 83-85 in RundeckKey.java already parse timestamps from the stored data).

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 2]
```diff
@@ -148,6 +153,29 @@ ResourceBase loadResource(){
...
 
+            // Parse and set timestamps from Vault metadata (KV v2)
+            if (this.vaultMetadata != null && !this.vaultMetadata.isEmpty()) {
+                String createdTime = this.vaultMetadata.get("created_time");
+
+                if (createdTime != null && !createdTime.isEmpty()) {
+                    try {
+                        // Vault returns timestamps in RFC3339 format (ISO 8601)
+                        // Example: "2025-03-13T16:25:00.123456Z"
+                        SimpleDateFormat vaultDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH);
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 3]
```diff
@@ -14,7 +14,12 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.UnsupportedEncodingException;
+import java.text.DateFormat;
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 4]
```diff
@@ -148,6 +153,29 @@ ResourceBase loadResource(){
...
+
+                        builder.setCreationTime(creationDate);
+
+                        // Use created_time as modification time if no separate updated_time exists
+                        // In Vault KV v2, there's no separate "updated_time" in metadata,
+                        // so we use created_time for both
+                        builder.setModificationTime(creationDate);
+                    } catch (ParseException e) {
+                        // If parsing fails, timestamps will not be set and will default to relative time
+                    }
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
+package io.github.valfadeev.rundeck.plugin.vault;
+
+import io.github.jopenlibs.vault.response.LogicalResponse;
+import com.dtolabs.rundeck.core.storage.ResourceMeta;
+import org.junit.Test;
+import org.rundeck.storage.api.ContentMeta;
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java]: Solved

File: src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
+package io.github.valfadeev.rundeck.plugin.vault;
+
+import io.github.jopenlibs.vault.response.LogicalResponse;
+import com.dtolabs.rundeck.core.storage.ResourceMeta;
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 7]
```diff
@@ -148,6 +153,29 @@ ResourceBase loadResource(){
...
+                        Date creationDate = vaultDateFormat.parse(createdTime.substring(0, Math.min(19, createdTime.length())));
+
+                        builder.setCreationTime(creationDate);
+
+                        // Use created_time as modification time if no separate updated_time exists
+                        // In Vault KV v2, there's no separate "updated_time" in metadata,
+                        // so we use created_time for both
+                        builder.setModificationTime(creationDate);
+                    } catch (ParseException e) {
+                        // If parsing fails, timestamps will not be set and will default to relative time
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 8]
```diff
@@ -148,6 +153,29 @@ ResourceBase loadResource(){
...
+                        // Example: "2025-03-13T16:25:00.123456Z"
+                        SimpleDateFormat vaultDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH);
+                        Date creationDate = vaultDateFormat.parse(createdTime.substring(0, Math.min(19, createdTime.length())));
+
+                        builder.setCreationTime(creationDate);
+
+                        // Use created_time as modification time if no separate updated_time exists
+                        // In Vault KV v2, there's no separate "updated_time" in metadata,
+                        // so we use created_time for both
+                        builder.setModificationTime(creationDate);
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java [hunk 9]
```diff
@@ -45,10 +47,18 @@ KeyObject build(){
...
 
+            // Extract metadata from response for KV v2
+            Map<String, String> metadata = null;
+            DataMetadata dataMetadata = response.getDataMetadata();
+            if (dataMetadata != null && !dataMetadata.isEmpty()) {
+                metadata = dataMetadata.getMetadataMap();
+            }
+
             if(data !=null) {
                 object = new RundeckKey(response,path);
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java:59]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java [hunk 10]
```diff
@@ -97,6 +107,12 @@ public KeyObject getVaultParentObject(Path path){
...
             parentObject=new VaultKey(response, parentPath);
+
+            // Extract metadata for parent object too
+            DataMetadata dataMetadata = response.getDataMetadata();
+            if (dataMetadata != null && !dataMetadata.isEmpty()) {
+                parentObject.setVaultMetadata(dataMetadata.getMetadataMap());
+            }
         } catch (VaultException e) {
 
         }
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/KeyObjectBuilder.java:120]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 11]
```diff
@@ -148,6 +158,40 @@ ResourceBase loadResource(){
                 builder.setMeta(VaultStoragePlugin.RUNDECK_DATA_TYPE, "password");
             }
 
+            // Parse and set timestamps from Vault metadata (KV v2)
+            if (this.vaultMetadata != null && !this.vaultMetadata.isEmpty()) {
+                String createdTime = this.vaultMetadata.get("created_time");
+                String updatedTime = this.vaultMetadata.get("updated_time");
+
+                // Parse creation time
+                if (createdTime != null && createdTime.length() >= 19) {
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: Solved

File: src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+/**
+ * Unit tests for VaultKey timestamp handling
+ */
+public class VaultKeyTest {
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java:22]: Solved

File: src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+            vaultKey.setVaultMetadata(metadata);
+
+            // When: Loading the resource
+            ResourceBase resource = vaultKey.loadResource();
+
+            // Then: Resource should be created successfully
+            assertNotNull("Resource should not be null for timestamp: " + timestamp, resource);
+            assertNotNull("Resource contents should not be null for timestamp: " + timestamp,
+                         resource.getContents());
+        }
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java:190]: Solved

File: src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        // Then: Resource should be successfully created with metadata
+        assertNotNull("Resource should not be null", resource);
+        assertNotNull("Resource contents should not be null", resource.getContents());
+
+        // Verify the metadata map was used (the key will have been processed)
+        assertNotNull("Vault metadata should be set", vaultKey.getVaultMetadata());
+        assertEquals("Vault metadata should contain created_time",
+                    "2025-03-13T16:25:00.123456Z",
+                    vaultKey.getVaultMetadata().get("created_time"));
+    }
```
[the remark this answers is not in the document]
Comment (edbaltra) [reply src/test/java/io/github/valfadeev/rundeck/plugin/vault/VaultKeyTest.java:53]: Solved

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java [hunk 15]
```diff
@@ -148,6 +159,40 @@ ResourceBase loadResource(){
...
+                String updatedTime = this.vaultMetadata.get("updated_time");
+
+                // Parse creation time
+                if (createdTime != null && createdTime.length() >= RFC3339_DATETIME_PREFIX_LENGTH) {
+                    try {
+                        // Vault returns timestamps in RFC3339 format (ISO 8601) in UTC
+                        // Example: "2025-03-13T16:25:00.123456Z"
+                        SimpleDateFormat vaultDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH);
+                        vaultDateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+                        Date creationDate = vaultDateFormat.parse(createdTime.substring(0, RFC3339_DATETIME_PREFIX_LENGTH));
```
Comment (ronaveva) [inline src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: why are we getting just a substring of a date when trying to parse it?

**Comment by ronaveva** [comment]
it would be good to know what happens if the vault is not using the default date format

**Comment by edbaltra** [comment]
> it would be good to know what happens if the vault is not using the default date format

Latest commit adds a more robust a approach in case Vault change the default format in the future 👍

File: src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java (hunk 15 above)
Comment (edbaltra) [reply src/main/java/io/github/valfadeev/rundeck/plugin/vault/VaultKey.java]: The SimpleDateFormat pattern: "yyyy-MM-dd'T'HH:mm:ss" can only parse up to seconds, but with the latest commit that's not a problem anymore 👍

**Review by ronaveva — APPROVED** [review]
LGTM
